metrics/metrics.go

```diff
@@ -0,0 +1,220 @@
+//
+// See the [histogram docs] for more information
+//
+// [histogram docs]: https://prometheus.io/docs/tutorials/understanding_metric_types/#histogram
+func AddHistogram(key string, help string, labels []string) {
+	_, ok := counters[key]
+	if ok {
+		return
+	}
+	newHistogram := &histogramMetric{
+		key:    key,
+		labels: labels,
+		histogram: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    key,
+				Help:    help,
+				Buckets: defaultBuckets,
+			},
+			labels,
+		),
+	}
+	histograms[key] = newHistogram
+	registry.MustRegister(newHistogram.histogram)
+}
+
+// IncrementCounter will increment the counter specified by key by 1.
+// The labels provided by the labels property will be compared to the list of labels used to create the counter.
+// Any labels _not_ included in this call that were specified when the counter was created
+// will be given blank values (empty strings).
+// Any labels included in this call which were _not_ included when the counter was created
+// will be silently ignored
+func IncrementCounter(key string, labels map[string]string) {
+	counter := getCounter(key)
+	if counter == nil {
+		return
+	}
+	counter.With(labels).Inc()
+}
+
+// UpdateHistogram will update the histogram specified by key with the duration from howLong as Seconds.
+// So a duration of 500ms will be recorded as 0.5 in the histogram.
+// The labels provided by the labels property will be compared to the list of labels used to create the histogram.
+// Any labels _not_ included in this call there were specified when the histogram was created
+// will be given blank values (empty strings).
+// Any labels included in this call which were _not_ included when the histogram was created
+// will be silently ignored
+func UpdateHistogram(key string, howLong time.Duration, labels map[string]string) {
+	histogram := getHistogram(key)
+	if histogram == nil {
+		return
+	}
+	histogram.With(labels).Observe(howLong.Seconds())
+}
+
+func init() {
+	AddCounter(totalRequestsKey, "Total number of HTTP requests", []string{statusLabel, routeLabel})
+	AddCounter(totalErrorsKey, "Total number of errors (a status other than 2XX)", []string{statusLabel, routeLabel})
+	AddHistogram(requestDurationKey, "Duration of all HTTP requests", []string{statusLabel, routeLabel})
+}
+
+// GinMiddleware is a [gin.HandlerFunc] which will update the default http_requests_total, http_errors_total,
+// and http_request_duration_seconds for each request.
+// To include it in your service, just pass it to the `Use` function like so:
+//
+//	r := gin.Default()
+//	r.Use(metrics.GinMiddleware)
+func GinMiddleware(c *gin.Context) {
+	path := c.Request.URL.Path
+	s := time.Now()
+	c.Next()
+	e := time.Since(s)
+	status := c.Writer.Status()
+	statusLabels := map[string]string{statusLabel: fmt.Sprintf("%d", status), routeLabel: path}
+	UpdateHistogram(requestDurationKey, e, statusLabels)
+	IncrementCounter(totalRequestsKey, statusLabels)
+	if status < 200 || status > 299 {
```

Comment:
Do we consider 400, 404 etc error counts?  Bad request, bad paths etc... are not really error on Sprinkler.

Reply:
@schowsf think these would fit the generic `http_errors_total` definition, we can probably create a custom metric to capture errors we really care about for Sprinkler